### DISCUSSION TITLE
focus is not set to yes-button in modal dialog

### DISCUSSION THREAD

**Discussion by hgfrey** [primefaces/community#1744] [PrimeFaces] [answered]
Since PF 13.0.6, setting the focus with the 'focus' attribute in the 'dialog' no longer works. The element briefly gains focus and then immediately loses it again. This still worked in PF 13.0.5.

Is this known? Should I open an issue for this?

**Comment by hgfrey** [comment]
Since PF 13.0.6 active element jumps back to the link I have clicked.

**Reply by melloware** [reply]
Its working fine for me:
[discussion-1744.zip](https://github.com/primefaces/community/files/15082144/discussion-1744.zip)

Run `mvn clean jetty:run` and navigate to http://localhost:8080/

When you open the dialog the `"Yes"` button is focus and when you close the dialog it returns focus back to the original calling button `"Show"`.

**Reply by hgfrey** [reply]
I will try to build a PF-Test-Project.

**Comment by hgfrey** [comment]
I've had this problem before:

https://github.com/orgs/primefaces/discussions/204

If I increase the timeout from 100 to 150 milliseconds, the focus remains on the dialog or is set here again.

I tried to build a tester to display the problem. But I didn't succeed. The dialogue works as expected.

Something is taking my focus away, but what?

**Reply by melloware** [reply]
Something in your code. You have something custom going on

**Reply by hgfrey** [reply]
Setting the focus no longer works for us since we switched from version PF 11 to 13. This change has a separate commit. The commit before works and the one with the change no longer works. We skipped version 12. But even with this version the focus still remains where it should be. Only with version 13 does the focus jump back to the triggering link.

**Reply by melloware** [reply]
Can you reproduce it with a PrimeFaces Test?  As you can see mine is using PF 13 and working perfectly under ARIA and Accessibility guidelines focusing the correct Button and then returning focus to the originating caller.  If you have a PF Test reproducer I can take a look, if not you have to assume you have a bug in your application.

**Reply by hgfrey** [reply]
I'm trying to gradually incorporate our application into your tester. Let's see what happens.

**Reply by melloware** [reply]
Excellent. That is the only way to truly get to the bottom of a problem.

**Reply by hgfrey** [reply]
Unfortunately, the next attempt also works. :-(

**Reply by hgfrey** [reply]
If I set global="false" the application works as expected. But I still need the extra line from:

https://github.com/orgs/primefaces/discussions/204

PrimeFaces.current().executeScript("setTimeout( () => PF('" + widgetVar + "').applyFocus(), 100 )");

**Comment by hgfrey** [comment]
Unfortunately, the next attempt also works. :-(

**Comment by hgfrey** [comment]
I'll look at it again tomorrow.

**Reply by melloware** [reply]
OK just let us know...

**Reply by hgfrey** [reply]
Please have a look at my tester.

When I click on the button, setting the focus works, but when I click on one of the two links it doesn't work.

[primefaces-test-master.zip](https://github.com/primefaces/community/files/15108999/primefaces-test-master.zip)

**Reply by melloware** [reply]
Awesome!  You are not crazy.  I see it too so I am debugging to figure out why CommandLink is different!

**Reply by hgfrey** [reply]
Thank you

**Reply by melloware** [reply]
Ticket opened:  https://github.com/primefaces/primefaces/issues/11818

Fixed for 13.0.9!

**Reply by hgfrey** [reply]
super!

**Comment by hgfrey** [comment]
Unfortunately, the focus is still not set. Now it is no longer set when I press the button. The tester I have provided here shows that the focus is not set. Instead there is now an error on the console:

Uncaught TypeError: this.cfg is undefined
    applyFocus http://localhost:8080/primefaces-test/javax.faces.resource/components.js.xhtml?ln=primefaces&v=13.0.9:341
    setTimeout handler*applyFocus http://localhost:8080/primefaces-test/javax.faces.resource/components.js.xhtml?ln=primefaces&v=13.0.9:341
    postShow http://localhost:8080/primefaces-test/javax.faces.resource/components.js.xhtml?ln=primefaces&v=13.0.9:340
    _show http://localhost:8080/primefaces-test/javax.faces.resource/components.js.xhtml?ln=primefaces&v=13.0.9:340
    show http://localhost:8080/primefaces-test/javax.faces.resource/components.js.xhtml?ln=primefaces&v=13.0.9:339
    <anonymous> http://localhost:8080/primefaces-test/ line 2 > injectedScript:1
    jQuery 2
    eval http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:73
    doEval http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:67
    handle http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:64
    b http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:58
    jQuery 6
    send http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:56
    offer http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:45
    handle http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:47
    ab http://localhost:8080/primefaces-test/javax.faces.resource/core.js.xhtml?ln=primefaces&v=13.0.9:39
    onclick http://localhost:8080/primefaces-test/:1

**Reply by melloware** [reply]
Sorry you will need this MonkeyPatch.

```js
if (PrimeFaces.widget.Dialog) {
    PrimeFaces.widget.Dialog.prototype.applyFocus = function() {
        if (this.cfg.focus) {
            var $this = this;
            setTimeout(function() {
                PrimeFaces.expressions.SearchExpressionFacade.resolveComponentsAsSelector($this.cfg.focus).trigger('focus');
            }, 100);
        } else
            PrimeFaces.focus(null, this.id);
    }
}
```

**Comment by hgfrey** [comment]
The patch works for the commandLink, but not for my commandButton. I have to put together an example of this tomorrow. In my current tester both the link and the button work.

**Reply by melloware** [reply]
@hgfrey i am using the test code you sent and its working fine?  Attached.
[discussion-1744-2.zip](https://github.com/primefaces/community/files/15192142/discussion-1744-2.zip)

**Reply by hgfrey** [reply]
That's correct. I have to check this.

**Reply by hgfrey** [reply]
It doesn't work in our application either. It works when I use the workaround from a few weeks ago.

PrimeFaces.current().executeScript("setTimeout( () => PF('" + widgetVar + "').applyFocus(), 100 )");

Otherwise not.

**Reply by hgfrey** [reply]
BTW: Where should I place the patch? In maine JS file or directly in the xhtml page?

**Reply by melloware** [reply]
personally i would add it to a JS file that gets included LAST in your order.

**Reply by melloware** [reply]
Oh if your old code works then you have definitely not applied the patch correctly?

**Reply by hgfrey** [reply]
The patch works if I set the timout to 150 milliseconds.

Of course, it would be nicer if the element got and kept the focus directly. At 150 milliseconds I have a very noticeable flickering. Will there possibly be a “better” solution in a later version?

**Reply by melloware** [reply]
You are more than welcome to analyze the issue and provide a better fix.  Also it interesting you used 100 in your server side fix but 100ms is not enough on the client side?

The issue is until the dialog is rendered on screen and visible the. It can receive focus from the calling button or link. That is the need for the delay.

**Comment by martin654** [comment]
Hi, we are having issues with focus when opening dialog since PF 13.0.6. In PF 13.0.5 it worked smoothly.

I created reproducer - please see attached .zip project. It is a bit more complex - we use ajaxStatus and we block user interaction until ajax is completed.
I tried PF 13.0.5, PF 13.0.6 and 13.0.9. It works in 13.0.5 - focus is set on input inside dialog. 
I noticed a bit different behavior when dialog opened using mouse click on button vs default command from input in page.

Should I create another issue?

Thank you

[primefaces-test_dialog_focus.zip](https://github.com/primefaces/community/files/15196984/primefaces-test_dialog_focus.zip)

**Reply by melloware** [reply]
@martin654 can you check what changed between .05 and .06to narrow down which change it was?

The only one I really see is this one: https://github.com/primefaces/primefaces/issues/11318

And yes open a new issue with reproducer.

**Reply by martin654** [reply]
@melloware  created https://github.com/primefaces/primefaces/issues/11860

**Reply by melloware** [reply]
OK fixed it and added a MonkeyPatch in the ticket.  It was related to you using both an AjaxStatusDialog and regular dialog and timing of AJAX ordering etc.  Good reproducer!

**Reply by melloware** [reply]
@martin654 @hgfrey can you both try 13.0.10 i just released to Maven Central.  Remove your focus MonkeyPatches and let me know if it is working.  

@hgfrey i still don't understand your timeout at 150ms issue as you previously had it at 100ms and it was working and your reproducer you sent me is working at 100ms.

**Reply by hgfrey** [reply]
I removed the patch and integrated version 13.0.10, now the focus is set again as desired. Excellent! Good job. Thank you very much for the quick help.

**Reply by martin654** [reply, marked as answer]
@melloware Thank you for the fix, works great in 13.0.10.

**Reply by melloware** [reply]
Awesome thanks guys!!
